internal/event/event.go

```diff
@@ -85,28 +107,66 @@ func IsEventDiscarded(err error) bool {
 	return errors.Is(err, ErrEventDiscarded)
 }
 
+func IsEventNotAllowed(err error) bool {
+	return errors.Is(err, ErrEventNotAllowed)
+}
+
+func NewEventNotAllowedErr(format string, a ...any) error {
+	return fmt.Errorf("%w: %w", ErrEventNotAllowed, fmt.Errorf(format, a...))
+}
+
+func NewEventDiscardedErr(format string, a ...any) error {
+	return fmt.Errorf("%w: %w", ErrEventDiscarded, fmt.Errorf(format, a...))
+}
+
 func (evs EventSource) ApplicationEvent(evType EventType, app *v1alpha1.Application) *cloudevents.Event {
 	cev := cloudevents.NewEvent()
 	cev.SetSource(evs.source)
 	cev.SetSpecVersion(cloudEventSpecVersion)
 	cev.SetType(evType.String())
+	cev.SetExtension(eventID, createEventID(app.ObjectMeta))
+	cev.SetExtension(resourceID, createResourceID(app.ObjectMeta))
 	cev.SetDataSchema(TargetApplication.String())
 	// TODO: Handle this error situation?
 	_ = cev.SetData(cloudevents.ApplicationJSON, app)
 	return &cev
 }
 
+func createResourceID(res v1.ObjectMeta) string {
+	return fmt.Sprintf("%s_%s", res.Name, res.UID)
```

Comment:
Currently, in this PR, the UID is taken from `.metadata`. In order to ensure consistency between source of truth (e.g. principal) and peer (e.g. managed agent), we need to use the UID from the source of truth on both sides:
- Resources events are sent between principal <-> agent
- The UID of the resource (application/appproject) is now included as part of the event, as part of this PR.
- However, the uid value MUST be the same on both the principal and agent sides.
- Our logic would be: when the agent/principal sees an event for a resource (e.g. application) with a matching name, but without have a matching uid, our logic should be to assume that the object with the old uuid was deleted.
- At present, (correct me if wrong!) this PR is using `.metadata.UID` for both principal and agent sides. This is a problem (likely that we will hit later), as this value will be different on both sides.
    - The `.metadata.uid` value is generated by K8s when the object is created.
    - However this means it is generated twice: one on the principal side when the Application is created in the principal namespace, and again on the agent side, when the agent completes the Application sync from principal.
    - Since they are being created on separate clusters, and the UID value is not used on creation, the values will be different.
- The canonical UID for a resource should come from the source of truth for that resource:
    - For managed agent, the `.metadata.UID` on principal is the source of truth
    - For autonomous agent, the `.metadata.UID` on agent is the source of truth.
- So, we can still use `.metadata.UID` when sending events from source of truth (principal), but the non-source of truth peer needs a way to know what UUID to use.

### Proposed Solution:
- For this, I recommend we add an annotation like `application-uid: (uid from principal)` to the resource on the agent (e.g. non source of truth) side.
- Then, when the agent is processing an event, it can use this uuid judge whether or not the event from principal still applies (and vice versa)
    - If the UID matches, the event still applies. If the UID doesn't match, the event no longer applies.

Reply:
Currently, we use UIDs in this PR to handle ACKs. We add the `.metadata.UID` of the resources from the source of truth, and the same UID is used on the other side. For example, in the case of managed mode:
1. When an event occurs on the Principal, we attach a resource's `metadata.UID` to an event and send it to the other side.
2. The agent receives this event and processes the event without updating the UID. It sends back an ACK event with the same UID.
4. The principal receives the ACK and verifies if the UID matches.

I think we will still face issues when handling `delete` events and I like the annotation approach for verification. We already keep track of all the managed apps and their resourceVersions using [ApplicationManager](https://github.com/argoproj-labs/argocd-agent/blob/a959b0e4246ac428afd92a8159d00effb96cf27e/internal/manager/application/application.go#L58). Including UID there could be one more option. If it's okay can we address it in a separate PR?